consumer.go

```diff
@@ -89,298 +145,453 @@ func NewConsumer(client *Client, topic string, partition int32, group string, co
-		return true
+	return worker
+}
+
+func (c *Consumer) unrefWorker(broker *Broker) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	worker := c.workers[broker]
+	worker.refs--
+
+	if worker.refs == 0 {
+		close(worker.input)
+		delete(c.workers, broker)
 	}
 }
 
-func (c *Consumer) fetchMessages() {
+// PartitionConsumer
 
-	fetchSize := c.config.DefaultFetchSize
+// PartitionConsumer processes Kafka messages from a given topic and partition. You MUST call Close()
+// on a consumer to avoid leaks, it will not be garbage-collected automatically when it passes out of
+// scope (this is in addition to calling Close on the underlying consumer's client, which is still necessary).
+type PartitionConsumer struct {
+	consumer  *Consumer
+	config    PartitionConsumerConfig
+	topic     string
+	partition int32
 
-	for {
-		request := new(FetchRequest)
-		request.MinBytes = c.config.MinFetchSize
-		request.MaxWaitTime = int32(c.config.MaxWaitTime / time.Millisecond)
-		request.AddBlock(c.topic, c.partition, c.offset, fetchSize)
-
-		response, err := c.broker.Fetch(c.client.id, request)
-		switch {
-		case err == nil:
-			break
-		case err == EncodingError:
-			if c.sendError(err) {
-				continue
-			} else {
-				return
-			}
+	broker         *Broker
+	events         chan *ConsumerEvent
+	trigger, dying chan none
+
+	fetchSize int32
+	offset    int64
+}
+
+func (child *PartitionConsumer) sendError(err error) {
+	child.events <- &ConsumerEvent{
+		Topic:     child.topic,
+		Partition: child.partition,
+		Err:       err,
+	}
+}
+
+func (child *PartitionConsumer) dispatcher() {
+	for _ = range child.trigger {
+		select {
+		case <-child.dying:
+			close(child.trigger)
 		default:
-			Logger.Printf("Unexpected error processing FetchRequest; disconnecting broker %s: %s\n", c.broker.addr, err)
-			c.client.disconnectBroker(c.broker)
-			for c.broker, err = c.client.Leader(c.topic, c.partition); err != nil; c.broker, err = c.client.Leader(c.topic, c.partition) {
-				if !c.sendError(err) {
-					return
+			if child.broker != nil {
+				child.consumer.unrefWorker(child.broker)
+				child.broker = nil
+			}
+
+			if err := child.dispatch(); err != nil {
+				child.sendError(err)
+				child.trigger <- none{}
+
+				// there's no point in trying again *right* away
+				select {
+				case <-child.dying:
+					close(child.trigger)
+				case <-time.After(10 * time.Second):
 				}
 			}
-			continue
 		}
+	}
 
-		block := response.GetBlock(c.topic, c.partition)
-		if block == nil {
-			if c.sendError(IncompleteResponse) {
-				continue
-			} else {
-				return
-			}
+	if child.broker != nil {
+		child.consumer.unrefWorker(child.broker)
+	}
+	child.consumer.removeChild(child)
+	close(child.events)
+}
+
+func (child *PartitionConsumer) dispatch() error {
+	if err := child.consumer.client.RefreshTopicMetadata(child.topic); err != nil {
+		return err
+	}
+
+	if leader, err := child.consumer.client.Leader(child.topic, child.partition); err != nil {
```

Comment:
The problem is the `addChild()` in the middle of this code in `ConsumerPartition` - if I just call `dispatch()` then that call gets moved and things can happen in the wrong order. It's not obvious to me how to fix that without just duplicating the code.